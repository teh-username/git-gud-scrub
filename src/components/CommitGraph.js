import React from 'react';
import vis from 'vis';

const canvasId = 'commitGraph';
const visConfig = {
  nodes: {
    shape: 'circle',
  },
  physics: {
    enabled: false,
  },
  interaction: {
    dragNodes: false, // do not allow dragging nodes
    zoomView: true, // allow zooming
    dragView: true, // allow dragging
  },
  layout: {
    hierarchical: {
      direction: 'LR',
      sortMethod: 'directed',
    },
  },
};

class CommitGraph extends React.Component {
  constructor(props) {
    super(props);
    this._renderGraph = this._renderGraph.bind(this);
  }

  _renderGraph(graphData) {
    const { edges, nodes } = graphData;
    const container = document.getElementById(canvasId);
    const data = {
      nodes: new vis.DataSet(nodes),
      edges: new vis.DataSet(edges),
    };
    new vis.Network(container, data, visConfig);
  }

  componentDidMount() {
    this._renderGraph(this.props.graphData);
  }

  shouldComponentUpdate(nextProps) {
    this._renderGraph(nextProps.graphData);
    return false;
  }

  render() {
    return (
      <div className="col-lg-9 col-sm-12">
        <h4>Commit Graph</h4>
        <div id={canvasId} className="commit-graph" />
      </div>
    );
  }
}

export default CommitGraph;
