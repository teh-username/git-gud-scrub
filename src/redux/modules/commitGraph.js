import { combineReducers } from 'redux';
import {
  GIT_BRANCH_ADD,
  GIT_BRANCH_DELETE,
  GIT_BRANCH_CHECKOUT,
  GIT_COMMIT,
} from './gitEmulator';
import grapher from '../../utils/grapher';

const initialState = {
  head: 'qa9j',
  branches: {
    master: 'qa9j',
    derp: 'kgxu',
    nope: 'wnrw',
  },
  commits: ['ko9z', 'plve', 'cb54', 'wnrw', 'kgxu', 'qa9j'],
  lookup: {
    ko9z: {
      message: 'Lorem ipsum dolor sit amet',
      children: ['plve', 'cb54'],
    },
    plve: {
      message: 'consectetur adipiscing elit',
      children: ['wnrw'],
    },
    cb54: {
      message: 'Donec molestie dictum dolor',
      children: ['kgxu', 'qa9j'],
    },
    wnrw: {
      message: 'Curabitur maximus ante velit',
      children: [],
    },
    kgxu: {
      message: ' Aenean sit amet enim at tellus vulputate consectetur',
      children: [],
    },
    qa9j: {
      message: 'Nullam auctor sapien a laoreet luctus',
      children: [],
    },
  },
};

const lookup = (state = initialState.lookup, action) => {
  switch (action.type) {
    case GIT_COMMIT:
      return {
        ...state,
        [action.parentRef]: {
          ...state[action.parentRef],
          children: [...state[action.parentRef].children, action.ref],
        },
        [action.ref]: {
          message: action.message,
          children: [],
        },
      };
    default:
      return state;
  }
};
const commits = (state = initialState.commits, action) => {
  switch (action.type) {
    case GIT_COMMIT:
      return [...state, action.ref];
    default:
      return state;
  }
};
const branches = (state = initialState.branches, action) => {
  switch (action.type) {
    case GIT_BRANCH_ADD:
      return {
        ...state,
        [action.name]: action.ref,
      };
    case GIT_BRANCH_DELETE:
      return Object.entries(state)
        .filter(([branch, ref]) => branch !== action.name)
        .reduce(
          (acc, [branch, ref]) => ({
            ...acc,
            [branch]: ref,
          }),
          {}
        );
    case GIT_COMMIT:
      return Object.entries(state)
        .filter(([branch, ref]) => ref === action.parentRef)
        .reduce(
          (acc, [branch, _]) => ({
            ...acc,
            [branch]: action.ref,
          }),
          { ...state }
        );
    default:
      return state;
  }
};
const head = (state = initialState.head, action) => {
  switch (action.type) {
    case GIT_BRANCH_CHECKOUT:
    case GIT_COMMIT:
      return action.ref;
    default:
      return state;
  }
};

export default combineReducers({
  lookup,
  commits,
  branches,
  head,
});

export const getGraphFromState = ({
  commitGraph: { lookup, commits, branches, head },
}) =>
  grapher({
    lookup,
    commits,
    branches,
    head,
  });

export const getHeadReference = ({ commitGraph: { head } }) => head;
export const getBranchReference = ({ commitGraph: { branches } }, branch) =>
  branches[branch];
export const getBranches = ({ commitGraph: { branches } }) =>
  Object.keys(branches);
