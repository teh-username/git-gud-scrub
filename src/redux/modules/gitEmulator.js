export const GIT_ADD = 'modules/gitEmulator/GIT_ADD';

export const gitAdd = fileName => ({
  type: GIT_ADD,
  fileName
});

const gitCommandActionLookup = {
  add: gitAdd
};

export const executeCommand = input => dispatch => {
  // TODO: add check if command does not exist
  const { command, argument, flags } = input;
  const gitCommand = gitCommandActionLookup[command];
  dispatch(gitCommand(argument, flags));
};
