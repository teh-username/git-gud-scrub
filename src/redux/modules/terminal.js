import { combineReducers } from 'redux';

const ADD_LOG_INFO = 'modules/terminal/ADD_LOG_INFO';
export const ADD_LOG_ERROR = 'modules/terminal/ADD_LOG_ERROR';

/*
  Sample state:
  {
    entries: [
      {
        text: 'git status',
        logLevel: 'info',
      },
      {
        text: 'wat',
        logLevel: 'error',
      },
    ]
  }
*/
export const initialState = {
  entries: [],
};

export const entries = (state = initialState.entries, action) => {
  switch (action.type) {
    case ADD_LOG_INFO:
      return [...state, { text: action.text, logLevel: 'info' }];
    case ADD_LOG_ERROR:
      return [...state, { text: action.text, logLevel: 'error' }];
    default:
      return state;
  }
};

export default combineReducers({
  entries,
});

export const addLogInfo = text => ({
  type: ADD_LOG_INFO,
  text,
});

export const addLogError = text => ({
  type: ADD_LOG_ERROR,
  text,
});

export const getEntries = ({ terminal }) => terminal.entries;
