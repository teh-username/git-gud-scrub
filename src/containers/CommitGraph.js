import { connect } from 'react-redux';

import CommitGraph from '../components/CommitGraph';
import { getGraphFromState } from '../redux/modules/commitGraph';

const mapStateToProps = state => ({
  graphData: getGraphFromState(state),
});

export default connect(mapStateToProps)(CommitGraph);
