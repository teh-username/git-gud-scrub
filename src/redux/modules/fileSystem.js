import { combineReducers } from 'redux';
import { GIT_ADD, GIT_COMMIT } from './gitEmulator';

export const MAX_FILE_COUNT = 5;
const ADD_FILE = 'modules/fileSystem/ADD_FILE';
const MODIFY_FILE = 'modules/fileSystem/MODIFY_FILE';

/*
  Sample state:
  {
    files: ['woof.py', 'ruff.py'],
    fileStatus: {
      'woof.py': {
        tracked: true,
        modified: true,
        staged: true,
      },
      'ruff.py': {
        tracked: true,
        modified: true,
        staged: true,
      },
    },
  };
*/

export const initialState = {
  files: [],
  fileStatus: {},
};

export const files = (state = initialState.files, action) => {
  switch (action.type) {
    case ADD_FILE:
      return [...state, action.fileName];
    default:
      return state;
  }
};

export const fileStatus = (state = initialState.fileStatus, action) => {
  switch (action.type) {
    case ADD_FILE:
      return {
        ...state,
        [action.fileName]: {
          tracked: false,
          modified: undefined,
          staged: undefined,
        },
      };
    case MODIFY_FILE:
      return {
        ...state,
        [action.fileName]: {
          ...state[action.fileName],
          modified: true,
        },
      };
    case GIT_ADD:
      return {
        ...state,
        [action.fileName]: {
          ...state[action.fileName],
          tracked: true,
          staged: true,
          modified: false,
        },
      };
    case GIT_COMMIT:
      return Object.entries(state)
        .filter(([fileName, statuses]) => statuses.staged)
        .reduce((previousState, [fileName, statuses]) => {
          return {
            ...previousState,
            [fileName]: {
              ...previousState[fileName],
              staged: false,
            },
          };
        }, state);
    default:
      return state;
  }
};

export default combineReducers({
  files,
  fileStatus,
});

export const addFile = fileName => ({
  type: ADD_FILE,
  fileName,
});

export const modifyFile = fileName => ({
  type: MODIFY_FILE,
  fileName,
});

export const getFiles = ({ fileSystem }) => fileSystem.files;
export const getFileStatus = ({ fileSystem }) => fileSystem.fileStatus;
