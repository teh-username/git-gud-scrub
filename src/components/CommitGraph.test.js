import React from 'react';
import CommitGraph from './CommitGraph';
import { initialState, getGraphFromState } from '../redux/modules/commitGraph';

describe('CommitGraph Component', () => {
  it('renders correctly', () => {
    let wrapper = shallow(
      <CommitGraph
        graphData={getGraphFromState({
          commitGraph: initialState,
        })}
      />,
      {
        disableLifecycleMethods: true,
      }
    );
    expect(wrapper).toMatchSnapshot();
  });
});
