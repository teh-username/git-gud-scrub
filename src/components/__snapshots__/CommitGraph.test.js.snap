// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitGraph Component renders correctly 1`] = `
<div
  className="col-lg-9 col-sm-12"
>
  <h4>
    Commit Graph
  </h4>
  <div
    className="commit-graph"
    id="commitGraph"
  />
</div>
`;
